src/server.go

```diff
@@ -0,0 +1,112 @@
+package fzf
+
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+)
+
+const (
+	crlf           = "\r\n"
+	httpOk         = "HTTP/1.1 200 OK" + crlf
+	httpBadRequest = "HTTP/1.1 400 Bad Request" + crlf
+)
+
+func startHttpServer(port int, channel chan []*action) {
+	if port == 0 {
+		return
+	}
+
+	listener, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
+	if err != nil {
+		return
+	}
+
+	go func() {
+		for {
+			conn, err := listener.Accept()
+			if err != nil {
+				if errors.Is(err, net.ErrClosed) {
+					break
+				} else {
+					continue
+				}
+			}
+			conn.Write([]byte(handleHttpRequest(conn, channel)))
+			conn.Close()
+		}
+		listener.Close()
+	}()
+}
+
+// Here we are writing a simplistic HTTP server without using net/http
+// package to reduce the size of the binary.
+//
+// * No --listen:            2.8MB
+// * --listen with net/http: 5.7MB
+// * --listen w/o net/http:  3.3MB
+func handleHttpRequest(conn net.Conn, channel chan []*action) string {
+	line := 0
+	headerRead := false
+	contentLength := 0
+	body := ""
+	bad := func(message string) string {
+		message += "\n"
+		return httpBadRequest + fmt.Sprintf("Content-Length: %d%s", len(message), crlf+crlf+message)
+	}
+	scanner := bufio.NewScanner(conn)
+	scanner.Split(func(data []byte, atEOF bool) (int, []byte, error) {
+		found := bytes.Index(data, []byte(crlf))
+		if found >= 0 {
+			token := data[:found+len(crlf)]
+			return len(token), token, nil
+		}
+		if atEOF || len(body)+len(data) >= contentLength {
+			return 0, data, bufio.ErrFinalToken
+		}
+		return 0, nil, nil
+	})
+
+	for scanner.Scan() {
+		text := scanner.Text()
+		if line == 0 && !strings.HasPrefix(text, "POST / HTTP") {
+			return bad("invalid request method")
+		}
+		if text == crlf {
+			headerRead = true
+		}
+		if !headerRead {
+			pair := strings.SplitN(text, ":", 2)
+			if len(pair) == 2 && strings.ToLower(pair[0]) == "content-length" {
+				length, err := strconv.Atoi(strings.TrimSpace(pair[1]))
+				if err != nil {
+					return bad("invalid content length")
+				}
+				contentLength = length
+			}
+		} else if contentLength <= 0 {
+			break
+		} else {
+			body += text
```

Comment:
There is no length/limit checking. If a malicious client keeps on sending garbage text, it gets added to `body`, till the process runs out of memory. I don't see a use-case where one would use this feature where the client cannot be trusted. However if it is used, then this could cause a crash. I think `net/http` does some limit checking.

Also if not fully validating content-length, why bother parsing? Some clients may not send it.

Reply:
> There is no length/limit checking

It is done here:

https://github.com/junegunn/fzf/blob/fd1f7665a77f3bd062586f0f3b54d4ff9863dd4b/src/server.go#L70-L71

I have tested this with `curl`, and if we don't stop after getting more than the specified content-length, the loop endlessly hangs trying to read more from the client, and curl never gets the response.

> I don't see a use-case where one would use this feature where the client cannot be trusted.

That's the assumption. `fzf < /dev/random` will easly make fzf consume an infinite amount of memory.

Anyway, thanks for chiming in. Is there anything you propose to better the implementation?

Reply:
https://github.com/junegunn/fzf/commit/1a9761736ed2190aa7536d4b6754f617f46cfd9c is an attempt to make the code slightly more robust.

Reply:
```suggestion
                        if len(body) > contentLength {
                            return bad("too much content")
                        }
			body += text
```
Improvements looks good but above check should resolve the infinite memory case fully.

Reply:
I don't think we have any memory problem now. Can you write an example of a malicious TCP request that can cause a memory problem?